src/sinks/datadog/logs.rs

```diff
@@ -2,59 +2,123 @@ use crate::{
+use serde_json::json;
+use std::io::Write;
 
-#[derive(Deserialize, Serialize, Debug)]
+#[derive(Deserialize, Serialize, Debug, Clone)]
 #[serde(deny_unknown_fields)]
 pub struct DatadogLogsConfig {
-    endpoint: Option<UriSerde>,
+    endpoint: Option<String>,
     api_key: String,
     encoding: EncodingConfig<Encoding>,
-    tls: Option<TlsConfig>,
+
+    #[serde(default)]
+    compression: Compression,
+
+    compression_level: Option<u32>,
+
+    #[serde(default)]
+    batch: BatchConfig,
+
+    #[serde(default)]
+    request: TowerRequestConfig,
+}
+
+#[derive(Clone)]
+pub struct DatadogLogsJsonService {
+    config: DatadogLogsConfig,
+}
+
+#[derive(Clone)]
+pub struct DatadogLogsTextService {
+    config: DatadogLogsConfig,
 }
 
 inventory::submit! {
     SinkDescription::new_without_default::<DatadogLogsConfig>("datadog_logs")
 }
 
+impl DatadogLogsConfig {
+    fn get_endpoint(&self) -> &str {
+        match &self.endpoint {
+            Some(ref endpoint) => &endpoint,
+            None => "https://http-intake.logs.datadoghq.eu/v1/input",
+        }
+    }
+}
+
 #[typetag::serde(name = "datadog_logs")]
 impl SinkConfig for DatadogLogsConfig {
     fn build(&self, cx: SinkContext) -> crate::Result<(VectorSink, Healthcheck)> {
-        let (host, port) = if let Some(uri) = &self.endpoint {
-            let host = uri
-                .host()
-                .ok_or_else(|| "A host is required for endpoint".to_string())?;
-            let port = uri
-                .port_u16()
-                .ok_or_else(|| "A port is required for endpoint".to_string())?;
-
-            (host.to_string(), port)
-        } else {
-            ("intake.logs.datadoghq.com".to_string(), 10516)
-        };
+        let request_settings = self.request.unwrap_with(&TowerRequestConfig::default());
+        let client = HttpClient::new(cx.resolver(), None)?;
 
-        let tls_settings = MaybeTlsSettings::from_config(
-            &Some(self.tls.clone().unwrap_or_else(TlsConfig::enabled)),
-            false,
-        )?;
+        // Create a different sink depending on which encoding we have chosen.
+        // Json and text have different batching strategies and so each needs to be
+        // handled differently.
+        match self.encoding.codec {
+            Encoding::Json => {
+                let batch_settings = BatchSettings::default()
+                    .bytes(bytesize::kib(100u64))
+                    .timeout(1)
+                    .parse_config(self.batch)?;
 
-        let sink = TcpSink::new(host, port, cx.resolver(), tls_settings);
-        let healthcheck = sink.healthcheck();
+                let service = DatadogLogsJsonService {
+                    config: self.clone(),
+                };
+                let healthcheck = healthcheck(service.clone(), client.clone()).boxed();
+                let sink = BatchedHttpSink::new(
+                    service,
+                    JsonArrayBuffer::new(batch_settings.size),
```

Comment:
Would it make sense to do the indirection within `BatchedHttpSink` instead of outside? ie:
```
let sink = BatchedHttpSink::new(
    Box::new(service) as Box<dyn HttpSink>,
    Box::new(JsonArrayBuffer::new(batch_settings.size)) as Box<dyn Batch>,
```

Reply:
The issue I run into here is that `BatchedHttpSink` wants an `HttpSink` and not a `Box<HttpSink>`. Dereffing the Box means I have a `&dyn HttpSink`, which is also not what I want. I imagine if I did get around that it wouldn't be `Sized`, that could also cause further issues.

Unless you have any other ideas, it's possible I'm missing something?

I see ultimately `BatchedHttpSink` stores an `Arc<HttpSink>`, with a bit of rework we could perhaps add a new `new` method that allowed an `Arc` to be passed in instead. But I'm not convinced it would be worth it just for this one case.

Reply:
Monomorphisation to the rescue.. I can, however, extract most of the duplicated functionality out into a separate function that each branch can then call.

Reply:
No, no other ideas, was just curious about where best to do the boxing.